src/assets/assets.service.ts

```diff
@@ -0,0 +1,61 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at https://mozilla.org/MPL/2.0/. */
+import { Injectable, NotFoundException } from '@nestjs/common';
+import { BasePrismaClient } from '../prisma/types/base-prisma-client';
+import { CreateAssetOptions } from './interfaces/create-asset-options';
+import { Asset, Transaction } from '.prisma/client';
+
+@Injectable()
+export class AssetsService {
+  async findByIdentifierOrThrow(
+    identifier: string,
+    prisma: BasePrismaClient,
+  ): Promise<Asset> {
+    const record = await prisma.asset.findUnique({
+      where: {
+        identifier,
+      },
+    });
+    if (!record) {
+      throw new NotFoundException();
+    }
+    return record;
+  }
+
+  async upsert(
+    options: CreateAssetOptions,
+    transaction: Transaction,
+    prisma: BasePrismaClient,
+  ): Promise<Asset> {
+    return prisma.asset.upsert({
+      create: {
+        identifier: options.identifier,
+        metadata: options.metadata,
+        name: options.name,
+        owner: options.owner,
+        supply: BigInt(0),
+        created_transaction_id: transaction.id,
+      },
+      update: {},
+      where: {
+        identifier: options.identifier,
+      },
+    });
+  }
+
+  async updateSupply(
+    asset: Asset,
+    delta: bigint,
+    prisma: BasePrismaClient,
+  ): Promise<Asset> {
+    return prisma.asset.update({
+      data: {
+        supply: asset.supply + delta,
```

Comment:
Will we run into any edge cases here if multiple `updateSupply` are happening at once? We could potentially revert the first delta right? I wonder if it is better to get supply and update in single transaction

Reply:
This will all happen within a single transaction, which is why the `BasePrismaClient` is passed around in all these methods. They will be used from the loader in a subsequent PR.